membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -58,13 +59,75 @@ object AccountHelpers {
-    AuthAndBackendViaAuthLibAction.async { implicit request =>
+  /**
+   * fetched with /v1/subscription/{key}?charge-detail=current-segment which zeroes out all the non-active charges
+   *
+   * There are multiple scenarios
+   *   - period between acquisition date and fulfilment date => None which indicates cancel now
+   *     - usually contractEffectiveDate to customerAcceptanceDate, except in the case of Guardian Weekly+6for6
+   *       where customerAcceptanceDate indicates start of GW proper invoiced period instead of start of 6for6
+   *       invoiced period despite GW+6for6 being just a regular Subscription with multiple products.
+   *     - free trial, or user choose start date of first issue in the future (lead time)
+   *   - Subscription within invoiced period proper => Some(endOfLastInvoicePeriod)
+   *   - free product => None which indicates cancel now
+   *   - edge case of being on the first day of invoice period however bill run has not yet happened => ERROR
+   *   - Today is after end of last invoice date and bill run has already completed => ERROR
+   *
+   * @return
+   *   Right(None) indicates cancel now,
+   *   Right(Some("yyyy-mm-dd")) indicates cancel at end of last invoiced period
+   *   Left indicates error and MMA should not proceed with automatic cancelation
+   */
+  private def calculateEffectiveCancellationDate[P <: SubscriptionPlan.AnyPlan : SubPlanReads](
+    identityId: String,
+    subscriptionName: memsub.Subscription.Name,
+    tp: TouchpointComponents
+  ): EitherT[Future, String, Option[LocalDate]] = {
+    EitherT(OptionT(tp.subService.get[P](subscriptionName, isActiveToday = true)).fold(
+      zuoraSubscriptionWithCurrentSegment => {
+        val paidPlans =
+          zuoraSubscriptionWithCurrentSegment.plans.list.collect { case paidPlan: PaidSubscriptionPlan[_, _] => paidPlan }
+        val now = LocalTime.now()
+        val billRunHasAlreadyHappened = now.isAfter(LocalTime.parse("12:00"))
+
+        paidPlans match {
+          case paidPlan1 :: paidPlan2 :: _ => \/.left("Failed to determine specific single active paid rate plan charge")
+
+          case paidPlan :: Nil => // single rate plan charge identified
+            paidPlan.chargedThrough match {
+              case Some(endOfLastInvoicePeriod) =>
+                val endOfLastInvoiceDateIsBeforeToday = endOfLastInvoicePeriod.isBefore(LocalDate.now())
+                if (endOfLastInvoiceDateIsBeforeToday && billRunHasAlreadyHappened)
+                  \/.left("chargedThroughDate exists but seems out-of-date because bill run should have moved chargedThroughDate to next invoice period. Investigate ASAP!")
+                else
+                  \/.right(Some(endOfLastInvoicePeriod))
+              case None =>
+                if (paidPlan.start.equals(LocalDate.now()) && !billRunHasAlreadyHappened) // effectiveStartDate exists but not chargedThroughDate
```

Comment:
I think these references to current time are going to cause us problems. Perhaps pass current date-time in as an argument?